### ISSUE TITLE
Cleanup walking through the wmb list in pioc_write_darray

### ISSUE THREAD

**Issue by jayeshkrishna** [NCAR/ParallelIO#611]
The code in pioc_write_darray() to walk through the wmb list (to get
to the end) can probably be cleaned up/simplified.

Code refactoring only : No change in behavior intended

**Comment by edhartnett** [comment]
Are you referring to this:

```
        if (recordvar)
        {
            /* Moving to the end of the wmb linked list to add the
             * current variable. Why are we checking the ioid here?
             * Don't all variables in the wmb have to have the same
             * decomposition? So ioid will always be set to the same
             * thing, and does not need to be checked when moving to
             * the end of the list. ??? */
            while(wmb->next && wmb->ioid != ioid)
                if (wmb->next)
                    wmb = wmb->next;
#ifdef _PNETCDF
            /* Do we still need the commented code below? ??? */
            /* flush the previous record before starting a new one. this is collective */
            /*       if (vdesc->request != NULL && (vdesc->request[0] != NC_REQ_NULL) ||
                     (wmb->frame != NULL && vdesc->record != wmb->frame[0])){
                     needsflush = 2;  // flush to disk
                     } */
#endif
        }
        else
        {
            /* Move to end of list. */
            while(wmb->next && wmb->ioid != -(ioid))
                if (wmb->next)
                    wmb = wmb->next;
        }

```

If so, I agree this needs to be cleaned up.

One issue is the overloading of the meaning of ioid, whcih, when negative, means a non-record variable. The code could be simplified by adding a new field to wmb called recordvar, and setting it appropriately. Then we can take out the confusing use of negative ioid, and then take out the checks for ioid = -ioid.

And of course the list can be walked in the usual way with a for statement.

**Comment by jayeshkrishna** [comment]
Yes, we can cleanup the code in different ways. Let me take a look at it and post my suggestion.

**Comment by edhartnett** [comment]
We can cleanup the code in different ways.

One way I would like this code cleaned up is to resolve the issue of overloading the ioid by trying to include too much information. That is a change that I need to have made.

If you are not up for it, it is fine, but then please leave this code to me, or explain how you will simplify it in a better way.

Thanks.

**Comment by edhartnett** [comment]
Looks like I will be cleaning up the negative ioid thing as part of my work on PIOc_write_darray(). So you may want to hold off until those changes are complete.

**Comment by jayeshkrishna** [comment]
ok, please make sure that the cleanup goes in as a separate PR.

**Comment by edhartnett** [comment]
It will probably be a series of PRs. ;-) I have one up now which just adds some documentation and logging. I will be adding some tests for the internal functions. Of course I want to add all the tests before actually doing any refactoring. ;-)

I will include you as a reviewer on all PRs related to the darray code.

**Comment by jayeshkrishna** [comment]
Thanks
